openfasoc/generators/gdsfactory-gen/relational.py

```diff
@@ -0,0 +1,754 @@
+    try:
+        glayout_actions.remove(GlayoutAction)
+    except ValueError:
+        pass
+    return glayout_actions
+
+
+def parse_direction(direction: str) -> Union[str,None]:
+    """parses a string to see if it contains a valid direction are up/north/above, right/east, left/west, down/south/below
+    Args: 
+        direction (str)
+    Returns None if no direction found or an integer 0-3 (0=west,1=north,2=south,3=east)
+    """
+    direction = direction.lower().strip()
+    findany = lambda strlst : any(ind in direction for ind in strlst)
+    if findany(["left","west"]):
+        return 0
+    elif findany(["up","north","above"]):
+        return 1
+    elif findany(["right","east"]):
+        return 2
+    elif findany(["down","south","below"]):
+        return 3
+    else:
+        return None
+
+    
+
+
+class ParametersList:
+    """store function parameter information
+    self.params is a list with the following format:
+        [ {name, defaultvalue, type}, ... ]
+    if niether type nor default is specified, both fields contain inspect.Parameter.empty
+    self.func_name: str the name of the function
+    """
+    class noprintobj:
+        def __init__(self,data):
+            self.data = data
+        def __str__(self):
+            return str(self.data)
+        def __repr__(self):
+            return str(self.data)
+    
+    def __init__(self, func)->dict:
+        self.func_name = func.__name__
+        parameters = inspect.signature(func).parameters
+        self.params = list()
+        for param_name, param in parameters.items():
+            paraminfo = {"name":param_name, 'defaultvalue':param.default, 'type':param.annotation}
+            self.params.append(paraminfo)
+        # formulate the grammar using the parameters
+        self.__grammar = nltk.CFG.fromstring(self.__get_str_grammar())
+        self.__grammar_parser = nltk.ChartParser(self.__grammar)
+
+    def __iter__(self):
+        return iter(self.params)
+    def names(self):
+        return [param["name"] for param in self.params]
+    def find(self, param_name: str):
+        """look for param with name param_name. 
+        If found, return info, else return None"""
+        for param in self.params:
+            if param["name"]==param_name:
+                return param
+        return None
+    def __str__(self):
+        return str(self.params)
+    
+    def __construct_glayers(self) -> dict:
+        glayerids = ['met1', 'met2', 'met3', 'met4', 'met5', 'metal1', 'metal2', 'metal3', 'metal4', 'metal5', 'capmet', 'active', 'diffusion', 'tap', 'welltap', 'n+s/d', 'p+s/d', 'polysilicon', 'poly', 'p+', 'n+', 'deep nwell', 'pwell', 'nwell', 'dnwell']
+        glayerids += ["metal 1", "metal 2","metal 3","metal 4","metal 5","met 1","met 2","met 3","met 4","met 5"]
+        glayerids += ["via 1", "via 2","via 3","via 4","via1","via2","via3","via4","metal contact","mcon","metal con"]
+        glayers = dict()
+        for glayerid in glayerids:
```

Comment:
This for loop seems unnecessary. Would be simpler to just define a dict directly. Also, can this be in MappedPDK instead?

Reply:
yeah I am trying to add many different ways of calling the same glayer. I could directly define as dictionary but it would be longer